### PR TITLE
Implementation of Conditional Gradient Optimizer 

### PR DESCRIPTION
**Pull request by pkan2**
Hello, 

I have uploaded the implementation of Conditional Gradient optimizer into this directory. It is implementing the issue #468. 

 The contributors are Pengyu Kan and Vishnu sai rao suresh Lokhande, whose github name is lokhande-vishnu.

Sincerely, 

Pengyu Kan

### CODE REVIEW

File: tensorflow_addons/optimizers/conditional_gradient_test.py [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+                        "ConditionalGraident/var3")
+            self.assertEqual(3, len(optimizer_variables))
+        '''
+
+    @test_utils.run_in_graph_and_eager_modes(reset_test=True)
+    def testMinimizeSparseResourceVariable(self):
+        for dtype in [tf.half, tf.float32, tf.float64]:
+            # This test invokes the ResourceSparseApplyConditionalGradient
+            # operation, which did not have a registered GPU kernel as of
+            # April 2018.
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient_test.py]: Is the limitation still an issue? Or can it be placed on GPU now?

File: tensorflow_addons/optimizers/BUILD [hunk 2]
```diff
@@ -9,6 +9,7 @@ py_library(
         "lazy_adam.py",
         "moving_average.py",
         "weight_decay_optimizers.py",
+        "conditional_gradient.py",
```
[the remark this answers is not in the document]
Comment (pkan2) [reply tensorflow_addons/optimizers/BUILD]: Thank! I will fixed them!

File: tensorflow_addons/optimizers/conditional_gradient_test.py (hunk 1 above)
Comment (pkan2) [reply tensorflow_addons/optimizers/conditional_gradient_test.py]: I think you are right and it should not be an issue here any more. Thank you for pointing it out! I will fix this comment part.

**Comment by Squadrick** [comment]
Could you update the PR after formatting as mentioned [here.](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#development-tips)

**Comment by pkan2** [comment]
Hi @Squadrick! Thank you for giving me the link! 

It is really helpful. But I have some error on "clang-format" while running the "make code-format". So I just manually run the yapf to change the format. Wish that it can work. Thank you!

**Comment by facaiy** [comment]
Hi, pkan2, did you try: `bash tools/run_docker.sh -c 'make code-format'` ? Note that it requires [docker](https://www.docker.com/products/docker-desktop) installed.

**Comment by facaiy** [comment]
Seems that we forget to append a comma in the line end :-)
```
FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
12c12
<         "weight_decay_optimizers.py"
---
>         "weight_decay_optimizers.py",
Please fix manually or run buildifier <file> to auto-fix.
Bazel configuration format check fails.
```

**Comment by pkan2** [comment]
Hello @Squadrick  and @facaiy! Thank you for your feedbacks!

I think the reason for failing the GPU test is caused by similar issue in #347. I am trying to use the solution provided in the this issue into the "conditional_gradient_test.py" file to get rid of the issue of "tf.half" in "'ResourceScatterUpdate' OpKernel" for GPU device. And it seems that "tensorflow_addons.utils" does not have attribute of "'GpuSupportsHalfMatMulAndConv'" from the solution. So I will probably just delete this part and keep the rest from the solution. 

Thank you!

Sincerely, 

Pengyu Kan

**Comment by Squadrick** [comment]
Yeah, that sounds good. We'll update the test once #347 is resolved. Could you add a TODO regarding this in the code?

**Comment by pkan2** [comment]
Hello @Squadrick. I have added TODO into the comment part of the code into the latest commit. 

Thank you!

**Comment by pkan2** [comment]
Hello @Squadrick! I am wondering whether there is more review or part that I have to modify on the file, before it can be merged? Because there is no more update on the comments for a few days. Thank you!

**Comment by Squadrick** [comment]
@pkan2 Could you pull from `master` and update the PR? This is to merge #482

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    variable -= (1-learning_rate)
+        * (variable + lamda * gradient / frobenius_norm(gradient))
+    ```
+    """
+
+    def __init__(self,
+                 learning_rate,
+                 lamda,
+                 use_locking=False,
+                 name="ConditionalGradient"):
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: Add `**kwargs` and pass that to `super()`, take a look at `LazyAdam`'s constructor for an example.

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        super(ConditionalGradient, self).__init__(name=name)
+        self._set_hyper("learning_rate", learning_rate)
+        self._set_hyper("lamda", lamda)
+
+    def get_config(self):
+        config = {
+            'learning_rate': self._learning_rate,
+            'lamda': self._lamda,
+            'use_locking': self._use_locking
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: `self._use_locking` hasn't been assigned in `__init__`

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    def _prepare(self, var_list):
+        learning_rate = self.learning_rate
+        if callable(learning_rate):
+            learning_rate = learning_rate()
+        self._learning_rate_tensor = tf.convert_to_tensor(
+            learning_rate, name="learning_rate")
+        lamda = self.lamda
+        if callable(lamda):
+            lamda = lamda()
+        self._lamda_tensor = tf.convert_to_tensor(lamda, name="lamda")
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: Add a `return super(ConditionalGradient, self)._prepare(var_list)`

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        var_update_op = tf.raw_ops.AssignVariableOp(**var_update_kwargs)
+        return tf.group(var_update_op)
+
+    def _resource_apply_sparse(self, grad, var, indices):
+        def frobenius_norm(m):
+            return tf.reduce_sum(m**2)**0.5
+
+        norm = tf.convert_to_tensor(frobenius_norm(grad), name="norm")
+        norm = tf.dtypes.cast(norm, var.dtype.base_dtype)
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: You can pass `var.dtype.base_type` to `convert_to_tensor()` which has a `dtype` param.

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        if callable(lamda):
+            lamda = lamda()
+        self._lamda_tensor = tf.convert_to_tensor(lamda, name="lamda")
+
+    def _resource_apply_dense(self, grad, var):
+        def frobenius_norm(m):
+            return tf.math.reduce_sum(m**2)**0.5
+
+        norm = tf.convert_to_tensor(frobenius_norm(grad), name="norm")
+        norm = tf.dtypes.cast(norm, var.dtype.base_dtype)
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: ditto

**Comment by pkan2** [comment]
I am so sorry @Squadrick . I messed up my forked repository while I try to update it with adding in the new merge. I can create a new new Pull Request later.

**Comment by pkan2** [comment]
Hi! I am so sorry for requesting for some many reviewers. But I think I have reverted the commit I mistakenly made. Thank you!

Hi @Squadrick! I hope I am doing the correct step that you described. I am not too sure for it. 
 And I have made modification based on your previous comments. Thank you!

**Comment by pkan2** [comment]
Hello @Squadrick! Can you help me check out on whether there is more things I have to modify on? Thank you! 

Sincerely,

Pengyu Kan

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                  applying gradients.  Defaults to "ConditionalGradient"
+        """
+        super(ConditionalGradient, self).__init__(name=name, **kwargs)
+        self._set_hyper("learning_rate", learning_rate)
+        self._set_hyper("lamda", lamda)
+        self._set_hyper("use_locking", use_locking)
+
+    def get_config(self):
+        config = {
+            'learning_rate': self._learning_rate,
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: Use `'learning_rate': self._serialize_hyperparameter('learning_rate')`

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            'lamda': self._lamda,
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: ditto

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            'use_locking': self._use_locking
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: ditto

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        base_config = super(ConditionalGradient, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def _create_slots(self, var_list):
+        for v in var_list:
+            self.add_slot(v, "conditional_gradient")
+
+    def _prepare(self, var_list):
+        learning_rate = self.learning_rate
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: `self.learning_rate` isn't set anywhere. (same for `self.lamda`)

I might be mistaken, but using `self._set_hyper("learning_rate", learning_rate)` doesn't create a new attr `self.learning_rate`. 

Here's the definition of `_set_hyper`:
```python
  def _set_hyper(self, name, value):
    """set hyper `name` to value. value can be callable, tensor, numeric."""
    if isinstance(value, trackable.Trackable):
      self._track_trackable(value, name, overwrite=True)
    if name not in self._hyper:
      self._hyper[name] = value
    else:
      prev_value = self._hyper[name]
      if (callable(prev_value)
          or isinstance(prev_value,
                        (ops.Tensor, int, float,
                         learning_rate_schedule.LearningRateSchedule))
          or isinstance(value, learning_rate_schedule.LearningRateSchedule)):
        self._hyper[name] = value
      else:
        backend.set_value(self._hyper[name], value)
```

You should instead be using `self._get_hyper('learning_rate')`

**Comment by pkan2** [comment]
Hi @WindQAQ @facaiy ! Can you provide some reviews or suggestions on the modification for the code? Thanks!

Sincerely,

Pengyu Kan

**Comment by pkan2** [comment]
Hi! @WindQAQ @facaiy! Are there any more suggestions on the modifications I have to work on?

Thanks!

**Comment by facaiy** [comment]
Sorry for the delay, @pkan2 . I'll take a look today. Thanks for ping me, Dheeraj :-)

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    def __init__(self,
+                 learning_rate,
+                 Lambda,
+                 use_locking=False,
+                 name="ConditionalGradient",
+                 **kwargs):
+        """Construct a conditional gradient optimizer.
+            Args:
+            learning_rate: A `Tensor` or a floating point value.
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/conditional_gradient.py]: Fix format for docstring.
```
"""Construct a conditional gradient optimizer.

Args:
    learning_rate: ...
    Lambda: ...
"""
```

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            'use_locking': self._serialize_hyperparameter('use_locking')
+        }
+        base_config = super(ConditionalGradient, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def _create_slots(self, var_list):
+        for v in var_list:
+            self.add_slot(v, "conditional_gradient")
+
+    def _prepare(self, var_list):
```
Comment (facaiy) [inline tensorflow_addons/optimizers/conditional_gradient.py]: It seems that  Optimizer base class introduced some new changes, can we refactor our codes according to https://github.com/tensorflow/tensorflow/blob/1aa61f86eef01fd1db124ccb55f49dd914540e96/tensorflow/python/keras/optimizer_v2/gradient_descent.py#L102 ?

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                 learning_rate,
+                 Lambda,
+                 use_locking=False,
+                 name="ConditionalGradient",
+                 **kwargs):
+        """Construct a conditional gradient optimizer.
+            Args:
+            learning_rate: A `Tensor` or a floating point value.
+                           The learning rate.
+            Lambda: A `Tensor` or a floating point value. The constraint.
```
[the remark this answers is not in the document]
Comment (pkan2) [reply tensorflow_addons/optimizers/conditional_gradient.py]: Thanks for the suggestion! It is just trying to keep the same letter used for expression in the formula by the paper. Can I choose "l_ambda" maybe? Otherwise, I can change into using "lambda_". 

Thank you!

File: tensorflow_addons/optimizers/conditional_gradient.py (hunk 13 above)
Comment (pkan2) [reply tensorflow_addons/optimizers/conditional_gradient.py]: I modify the "prepare" method into "prepare_local" method as done in the provided link and updated the "resource_apply_dense" and "resource_apply_sparse" accordingly. 

Thanks!

**Comment by pkan2** [comment]
@Squadrick @facaiy  Thank you a lot!

File: tensorflow_addons/optimizers/conditional_gradient.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+    ```
+    variable -= (1-learning_rate)
+        * (variable + l_ambda * gradient / frobenius_norm(gradient))
+    ```
+    """
+
+    def __init__(self,
+                 learning_rate,
+                 l_ambda,
```
Comment (facaiy) [inline tensorflow_addons/optimizers/conditional_gradient.py]: `l_ambda` doesn't follow name convention. I suggest to use `constraint` or `lambda_`, and explain it in the docstring. `constraint` looks better for me.

**Comment by facaiy** [comment]
Apologized that I'm busy this week than I expected. I'm fine to merge it if @Squadrick approved, and I'll recheck the PR later when I have time. 😄 thank all

File: tensorflow_addons/optimizers/conditional_gradient.py (hunk 15 above)
Comment (pkan2) [reply tensorflow_addons/optimizers/conditional_gradient.py]: I change the variable name into "lambda_". Thank you!

**Comment by pkan2** [comment]
Hi @Squadrick  @facaiy!  I have changed the variable name of constraint into "lambda_". And I have updated the README file with adding co-contributor Vishnu Lokhande into it as maintainer too. 

Thank you!
